### ISSUE TITLE
Fail request on to big to fit UDP message

### ISSUE THREAD

**Issue by doino-gretchenliev** [folbricht/routedns#93]
When a DNS response is too big to fit the UDP message size, the `miekg/dns` is raising a warning message:
> dns: buffer size too small

but the request doesn't fail and so `No answer` is returned. It would be nice for the request to fail in order for `Fail-Rotate` type of groups to be able to fall back to a TCP server.
`miekg/dns` discussion: https://github.com/miekg/dns/issues/214
test URL: https://www.emag.bg/

**Comment by folbricht** [comment]
Do you have a repro for the issue, maybe a small routedns config? I wonder if there are other ways to deal with this perhaps, but I need to first understand where exactly it's failing in this use-case.

**Comment by cbuijs** [comment]
**@doino-gretchenliev:**
A `dig` for `www.emag.bg` just works here. UDP-Size is 1232 for the answer (22 answer records), no problem on my setup. 

It might be that the DNS server you forwarding too does not support EDNS0 UDP sizes larger than 512 bytes, and reverts to truncating and TCP being used and this kicks things off the wrong way.

Try using DoH or DoT (if available) instead of plain DNS and it might work better as well.

**@folbricht:**
Not sure why it should not work, as EDNS0 negotiate UDP window-size, so probably if it regnogiate (and perhaps fallback to TCP), it goes wrong after. The miekg/dns module support this as it seems, but might be broken due to negotiating issues. Falling back on a UDP-Size of 512 when things go weird is probably the best handling of things (maybe configurable).

**Comment by cbuijs** [comment]
**@doino-gretchenliev** 

Indeed, I get the `WARN[0003] dns: overflowing header size as well`.

To workaround, you can change in the `dns.toml` file, the `protocol` for `resolvers.encrypt.udp` from `udp` to `tcp`, and then everything works fine. So it looks like this:
```
  [resolvers.encrypt-udp]
  address = "<ip_of_the_encrypt_dns_server>:53"
  protocol = "tcp"
```
**@folbricht** Seems EDNS is not added to a plain resolver and wipes out the UDP-sizing stuff. It seems only between two routedns entities it happens.

**Comment by doino-gretchenliev** [comment]
@cbuijs Yes, I've already switched to TCP, but I still prefer to use UDP where it's possible.

**Comment by cbuijs** [comment]
**@doino-gretchenliev** 

Gotcha!

I also tried to add ECS to make sure an EDNS pseudo-record exists, but no luck. So we need the sharp-eye and fix from **@folbricht** I am affraid :-).

**Comment by folbricht** [comment]
Thanks for the repro guys, that really helped a lot. I believe I have a fix for it, but it depends on if my PR is accepted upstream or not. That determines what exactly the fix will look like.

**Comment by folbricht** [comment]
I put a quick fix on the `issue-93` branch. I'll move the fix to a better place in the code if the upstream PR is accepted, but for now this is equivalent.

**Comment by folbricht** [comment]
Fixed in #95
